File: backend/src/main/java/fit/nlu/tmdt/modules/booking/service/BookingService.java
package fit.nlu.tmdt.modules.booking.service;

import fit.nlu.tmdt.modules.booking.dto.request.CreateBookingRequest;
import fit.nlu.tmdt.modules.booking.dto.response.BookingResponse;
import fit.nlu.tmdt.modules.booking.dto.response.TimeSlotResponse;

import java.time.LocalDate;
import java.util.List;

/**
 * Booking Service Interface
 */
public interface BookingService {

    /**
     * Create booking
     */
    BookingResponse createBooking(CreateBookingRequest request, Long userId);

    /**
     * Get user's bookings
     */
    List<BookingResponse> getUserBookings(Long userId);

    /**
     * Get landlord's bookings
     */
    List<BookingResponse> getLandlordBookings(Long landlordId);

    /**
     * Get booking by ID
     */
    BookingResponse getBookingById(Long bookingId, Long userId);

    /**
     * Confirm booking (landlord)
     */
    void confirmBooking(Long bookingId, Long landlordId);

    /**
     * Cancel booking (user or landlord)
     */
    void cancelBooking(Long bookingId, Long userId, String reason);

    /**
     * Complete booking (landlord)
     */
    void completeBooking(Long bookingId, Long landlordId, String note);

    /**
     * Mark as no-show (landlord)
     */
    void markNoShow(Long bookingId, Long landlordId, String reason);

    /**
     * Get available time slots for a post on a date
     */
    TimeSlotResponse getAvailableSlots(Long postId, LocalDate date);

    /**
     * Get calendar view for landlord
     */
    List<BookingResponse> getLandlordCalendar(Long landlordId, LocalDate startDate, LocalDate endDate);
}

File: frontend/src/services/bookingService.ts
import api from './api';
import type { ApiResponse, BookingResponse } from '@/types';

export interface CreateBookingRequest {
  postId: number;
  bookingTime: string;
  guestCount: number;
  note?: string;
  voucherCode?: string;
}

export const bookingService = {
  /**
   * Create a new booking/appointment
   */
  createBooking: async (data: CreateBookingRequest): Promise<BookingResponse> => {
    const response = await api.post<ApiResponse<BookingResponse>>('/v1/bookings', data);
    return response.data.data!;
  },

  /**
   * Get current tenant's bookings
   */
  getTenantBookings: async (): Promise<BookingResponse[]> => {
    const response = await api.get<ApiResponse<BookingResponse[]>>('/v1/bookings');
    return response.data.data!;
  },

  /**
   * Get landlord's bookings (Calendar view)
   */
  getLandlordBookings: async (startDate: string, endDate: string): Promise<BookingResponse[]> => {
    const response = await api.get<ApiResponse<BookingResponse[]>>('/v1/bookings/calendar', {
      params: { startDate, endDate },
    });
    return response.data.data!;
  },

  /**
   * Confirm booking (landlord)
   */
  confirmBooking: async (id: number | string): Promise<void> => {
    await api.put(`/v1/bookings/${id}/confirm`);
  },

  /**
   * Cancel booking
   */
  cancelBooking: async (id: number | string, reason?: string): Promise<void> => {
    await api.put(`/v1/bookings/${id}/cancel`, { reason });
  },

  /**
   * Complete booking (landlord)
   */
  completeBooking: async (id: number | string, note?: string): Promise<void> => {
    await api.put(`/v1/bookings/${id}/complete`, { note });
  },

  /**
   * Mark as no-show (landlord)
   */
  markNoShow: async (id: number | string, reason?: string): Promise<void> => {
    await api.put(`/v1/bookings/${id}/no-show`, { reason });
  },

  /**
   * Get booking details
   */
  getBookingById: async (id: number | string): Promise<BookingResponse> => {
    const response = await api.get<ApiResponse<BookingResponse>>(`/v1/bookings/${id}`);
    return response.data.data!;
  },
};

export default bookingService;
